### ISSUE TITLE
New rule suggestion: complete destructuring of parameters

### ISSUE THREAD

**Issue by whatisaphone** [typescript-eslint/typescript-eslint#1328]
Consider this code:

```ts
function Button({ text }: {
  text: string;
  onClick: () => void;
}) {
  return <button>{text}</button>;
}
```

The `onClick` prop was accidentally omitted, and no current rule is able to catch it (that I know of). I suggest a rule, called maybe `fully-destructure-arguments`, which would trigger if a property is present on an object type, but is not destructured.

It (probably) shouldn't trigger for interfaces et al, only immediate inline object types. An interface may be used elsewhere, but an inline object type is only used in one place, so if it's destructured and is missing a prop, the forgotten prop is guaranteed never to be used.

Bikeshedding is welcome on the rule name or the suggestion in general.

**Comment by bsene** [comment]
This issue seems to addresses a narrow bug (forgotten props on inline object types only), has had no community traction or PRs in 7 years, and the adjacent prefer-destructuring rule is frozen. 

IMO Teams needing this can use named types with no-unused-vars or a project-local custom rule.
@
